Modules/_zstd/zdict.c

```diff
@@ -0,0 +1,296 @@
+
+    /* Release dict_content after Free ZSTD_CDict/ZSTD_DDict instances */
+    Py_XDECREF(self->dict_content);
+
+    PyTypeObject *tp = Py_TYPE(self);
+    tp->tp_free((PyObject*)self);
+    Py_DECREF(tp);
+}
+
+/*[clinic input]
+_zstd.ZstdDict.__init__
+
+    dict_content: object
+        A bytes-like object, dictionary's content.
+    is_raw: bool = False
+        This parameter is for advanced user. True means dict_content
+        argument is a "raw content" dictionary, free of any format
+        restriction. False means dict_content argument is an ordinary
+        zstd dictionary, was created by zstd functions, follow a
+        specified format.
+
+Represents a zstd dictionary, which can be used for compression/decompression.
+
+It's thread-safe, and can be shared by multiple ZstdCompressor /
+ZstdDecompressor objects.
+[clinic start generated code]*/
+
+static int
+_zstd_ZstdDict___init___impl(ZstdDict *self, PyObject *dict_content,
+                             int is_raw)
+/*[clinic end generated code: output=c5f5a0d8377d037c input=e6750f62a513b3ee]*/
+{
+    /* Only called once */
+    if (self->inited) {
+        PyErr_SetString(PyExc_RuntimeError, init_twice_msg);
+        return -1;
+    }
+    self->inited = 1;
+
+    /* Check dict_content's type */
+    self->dict_content = PyBytes_FromObject(dict_content);
+    if (self->dict_content == NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                        "dict_content argument should be bytes-like object.");
+        return -1;
+    }
+
+    /* Both ordinary dictionary and "raw content" dictionary should
+       at least 8 bytes */
+    if (Py_SIZE(self->dict_content) < 8) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Zstd dictionary content should at least 8 bytes.");
+        return -1;
+    }
+
+    /* Get dict_id, 0 means "raw content" dictionary. */
+    self->dict_id = ZSTD_getDictID_fromDict(PyBytes_AS_STRING(self->dict_content),
+                                            Py_SIZE(self->dict_content));
+
+    /* Check validity for ordinary dictionary */
+    if (!is_raw && self->dict_id == 0) {
+        char *msg = "The dict_content argument is not a valid zstd "
+                    "dictionary. The first 4 bytes of a valid zstd dictionary "
+                    "should be a magic number: b'\\x37\\xA4\\x30\\xEC'.\n"
+                    "If you are an advanced user, and can be sure that "
+                    "dict_content argument is a \"raw content\" zstd "
+                    "dictionary, set is_raw parameter to True.";
+        PyErr_SetString(PyExc_ValueError, msg);
+        return -1;
+    }
+
+    return 0;
+}
+
+
+/*[clinic input]
+_zstd.ZstdDict.__reduce__
+
+
+Intentionally not supporting pickle.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd_ZstdDict___reduce___impl(ZstdDict *self)
+/*[clinic end generated code: output=5c9b8a3550429417 input=2d69a2b4775ff76d]*/
+{
+    PyErr_SetString(PyExc_TypeError,
```

Comment:
I think it's better if we document this instead. If you still want it, shouldn't we prefer `NotImplementedError` over `TypeError`?

Reply:
I agree documenting this instead is probably the better solution.